### PR TITLE
Make max_retries a string to allow variable interpolation

### PR DESCRIPTION
**Pull request by sylviamoss**
This is to allow variable interpolation in the `max_retries` option for JSON and HCL2. For backward compatibility, I've changed the provisioners decode to a weak decode. 

Closes https://github.com/hashicorp/packer/issues/9651

### CODE REVIEW

File: hcl2template/types.build.provisioners.go [hunk 1]
```diff
@@ -80,7 +80,7 @@ func (p *Parser) decodeProvisioner(block *hcl.Block) (*ProvisionerBlock, hcl.Dia
 		Except      []string `hcl:"except,optional"`
 		Rest        hcl.Body `hcl:",remain"`
 	}
-	diags := gohcl.DecodeBody(block.Body, nil, &b)
+	diags := gohcl.DecodeBody(block.Body, cfg.EvalContext(nil), &b)
```
Comment (sylviamoss) [inline hcl2template/types.build.provisioners.go:83]: this is to decode variables used as value like: 
```
max_retries = var.max_retries
```
if we don't decode at this point this kind of usage doesn't work

**Comment by SwampDragons** [comment]
I'm a little nervous about turning all decodes for the provisioners into a weak decode, but I don't have a valid reason other than "my gut says this will bite us in the future"

**Comment by sylviamoss** [comment]
We can remove the weak decode and release this on 1.7.0 with backward incompatibility. The weak decode is just to make it possible to use integer like before.

**Comment by azr** [comment]
Nice !

**Pull request merged**